### ISSUE TITLE
Scorecard - GitHub bot account

### ISSUE THREAD

**Issue by naveensrinivasan** [ossf/scorecard#1729]
**Is your feature request related to a problem? Please describe.**
In the process of automating scorecard-actions `e2e` testing a GitHub `token` is needed for creating issues on the scorecard-actions repo when these actions fail.

These actions run on a `cron-job`. Here is an example https://github.com/ossf-tests/scorecard-action-non-main-branch/tree/other

I am using my own personal token to run these actions and create an issue when these actions fail https://github.com/ossf/scorecard-action/issues/138

I wouldn't want my personal token to do this. 

The right approach would be to use `scorecard-bot` to do these jobs. 

@ossf/scorecard-admins Thoughts?

**Comment by laurentsimon** [comment]
DO you have documentation links about GitHub bot accounts, how to create them, etc?

**Comment by naveensrinivasan** [comment]
It is nothing but another GitHub account. The management of that account is the issue, who has the passwords for the account. Then the management of the keys that the account generates for automation etc.

@justaugustus Any thoughts on this?

**Comment by justaugustus** [comment]
> It is nothing but another GitHub account. The management of that account is the issue, who has the passwords for the account. Then the management of the keys that the account generates for automation etc.
> 
> @justaugustus Any thoughts on this?

@strawberry-baked-alaska @brianbehlendorf -- Can you look into [1password for open source teams](https://github.com/1Password/1password-teams-open-source) for @ossf/scorecard-admins?

I know CNCF (cc: @caniszczyk @amye) provides this and can be requested via CNCF Service Desk (I have a few vaults for Kubernetes SIG Release).

w.r.t. bots, ideally this is eventually a GitHub App (with real boteyness), but most/all of the "bots" I deal with today are shared GH accounts with the team holding creds in a vault somewhere.

I've squatted on @ossf-scorecard-bot, so we can use that (if everyone's in agreement).

**Comment by justaugustus** [comment]
My really rough cut of bot guidelines:
- identifiable name
- strong password (stored in a cred vault)
- 2FA enabled to hardware key (recovery codes stored in a cred vault)
- tightly-scoped PATs per purpose

**Comment by jorydotcom** [comment]
> @strawberry-baked-alaska @brianbehlendorf -- Can you look into [1password for open source teams](https://github.com/1Password/1password-teams-open-source) for @ossf/scorecard-admins?

we're on it!

**Comment by naveensrinivasan** [comment]
> > @strawberry-baked-alaska @brianbehlendorf -- Can you look into [1password for open source teams](https://github.com/1Password/1password-teams-open-source) for @ossf/scorecard-admins?
> 
> we're on it!

Great! Thanks

**Comment by laurentsimon** [comment]
just to clarify: the only permission is to be able to create issues, correct? And there's no dangerous permissions needed to do that, any PAT can do it. Is that correct?

**Comment by naveensrinivasan** [comment]
The bot can be extended for all the things like merging code, create issues, run e2e test etc. 

At the moment it is only for creating issues and run e2e.

**Comment by laurentsimon** [comment]
everything *but* merging code requires no dangerous permissions w.r.t the scorecard repo itself. I think what I was trying to gauge is how sensitive this bot account it. My understanding of "merging code" here is that it needs to bypass branch protection. It'd be nice to avoid doing this if we can

**Comment by naveensrinivasan** [comment]
> everything _but_ merging code requires no dangerous permissions w.r.t the scorecard repo itself. I think what I was trying to gauge is how sensitive this bot account it. My understanding of "merging code" here is that it needs to bypass branch protection. It'd be nice to avoid doing this if we can

I understand your concern. Merging code isn't required now. But when we plan for something like this it is good to plan for the long term.  The projects like `k8s`, `k8ssigs`, and other large GitHub repositories use `bot` to merge code. So we aren't going to be the first one.  

Also `merging code` shouldn't require "bypass branch" protection. Have you read this anywhere? If so can you please point to the doc?

**Comment by laurentsimon** [comment]
auto merging requires adding a review automatically. (added this to the auto-merge PR https://github.com/ossf/scorecard/pull/1749)

It also commented that we should explore "batched" dependabot PR, like once a week or something, which would avoid having an auto-merge feature

**Comment by justaugustus** [comment]
> we're on it!

Thanks Jory!!

---

I'll drop a few thoughts on things I'd like to see/have seen...

### Multiple roles for a bot

We have multiple bots in Kubernetes orgs.
Most/all of them are user accounts; most/all run do their tasks with as few privileges on the org as possible:

```yaml
  bots:
    description: Bot Service Accounts in the Kubernetes org
    maintainers:
    - k8s-ci-robot
    - k8s-github-robot
    - thelinuxfoundation
    members:
    - k8s-publishing-bot
    - k8s-release-robot
    privacy: closed
```

ref: https://github.com/kubernetes/org/blob/7b31d6ee782c82f5b44ec27600898271c292dbb4/config/kubernetes/org.yaml#L1425-L1434

In the case of `k8s-release-robot`, the account has multiple PATs; some have write access to repos, but most do not.
Setting the correctly-scoped tokens as organization secrets (`SCORECARD_BOT_READ|WRITE_TOKEN`), so we can leverage on any scorecard repo should solve the problem.

### Batching PRs / dependabot

On the roadmap, but no timeline as of yet.

ref: https://twitter.com/stephenaugustus/status/1504611650443390983, https://github.com/github/roadmap/issues/148, https://github.com/dependabot/dependabot-core/issues/1190

Batching is ideal, but approval is paramount.
Any PR that merges should be subject to same review/checks, both from maintainers and presubmits.

An additional gate that we have in Kubernetes is requiring an `ok-to-test` label on PRs from "non-trusted" (non-org members) before we run presubmits.

Because dependabot is a GitHub App (and not an org member), we get around this limitation in SIG Release repos by assuming dependabot is trusted, and having it automatically apply the `ok-to-test` label for PRs: https://github.com/kubernetes-sigs/release-utils/pull/38

Going a layer deeper on trust, certain repos/workflows can apply a `skip-review` label: https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.md#labels-that-apply-to-kubernetestest-infra-only-for-prs

This allows us to do things like automatically updating Prow images: https://github.com/kubernetes/test-infra/pull/25684
Motivations and some detail on workflow for that here: https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/2539-continuously-deploy-k8s-prow

Back to dependabot...

In instances that our tests match our expectations, then a test failure says, "I, as a maintainer, do not trust this code to be accepted".
This is currently true for any PR submitted.

For dependency updates, the human element of clicking `Approve` is mechanically and arguably, could be automated away.

Example workflow:
- @ossf-scorecard-bot is onboarded as a @ossf/scorecard-maintainers 
- bot scans for open PRs from dependabot that are not approved: https://github.com/ossf/scorecard/pulls?q=is%3Aopen+is%3Apr+author%3Aapp%2Fdependabot+review%3Arequired
- bot approves the PR
- --> auto-merge

Back to batching...

One of the things I crave in non-Kubernetes orgs is `tide` (the merge pool): https://github.com/kubernetes/test-infra/tree/master/prow/tide

We have a few expectations that are relevant here and enforced by branch protection:
- PR branches are up-to-date
- no merge commits / require linear history

We violate merge commits on pull requests by merging `main` in to the PR branch, but that violation is nullified by the fact that we prefer squash and merge as a merge strategy.

A little out-of-scope here, but I have [Opinions ™️](https://twitter.com/stephenaugustus/status/1488662159869501442) and did a Twitter poll to check-in with other maintainers on the internet if you want to read more: https://twitter.com/stephenaugustus/status/1488624720400502790

tl;dr snippet:

> My answer is "it depends".
>
> kubernetes/kubernetes does merge commits.
> I've found with small projects/fewer maintainers, squashing works really nicely, especially if it's solving/circumventing teaching people to rebase.
>
> As a more experienced maintainer, most of my commits are concise, focused on a specific area, and may include descriptions in the commit descriptions.
>
> For a larger PR, you're going to lose the value of being able to interrogate those structured commits.
>
> For a newer contributor, you may have a few scenarios:
> - poorly-structured commit history
> - commits which change multiple areas
> - someone who does not yet know how to manage that proactively
>
> Education is important here if you want to keep them contributing.

So in repos where multiple merge strategies are enabled, my preference is usually to do the following:

- `create a merge commit` if the submitter tells a good story with the commits and there's a higher value in having the individual commits stand on their own
- `sqaush and merge` when the commit history is messy

I mention merge strategy because every dependabot PR leads to churn in multiple human approvals and in instances where the go.sum has been changed in a way that triggers another dependabot to start rebasing itself.

This goes away with a merge queue.

I haven't played around with it yet, but I would suggest a tool like Mergify: https://mergify.com/ (which is free for open source projects)

I've enabled it on a few of my orgs just now, but would be good to get some opinions from maintainers here.

**Comment by laurentsimon** [comment]
> Batching is ideal, but approval is paramount. Any PR that merges should be subject to same review/checks, both from maintainers and presubmits.

> For dependency updates, the human element of clicking `Approve` is mechanically and arguably, could be automated away.

Human approval may also include taking a peek at the PR.
Any thoughts on combining dependabot PRs thru `git pull origin branch1 branch2`, to create a batched PR?
I've tested this idea in tests https://github.com/laurentsimon/scorecard-action-test-2/pull/19
The (draft) script https://github.com/laurentsimon/scorecard-action-test-2/blob/main/.github/workflows/combine-prs.yml iterates thru PRs, checks they are authored from dependabot, combines them using `git pull origin branch1 branch2 ...`, create a new PR, then closes the original ones.
The downside is that it creates a bit of noise due to PRs being closed by the bot.
I'm not 100% sure how reliable the combine part is, but I suspect it may work if the dependencies PR have no "conflicts". This does not require auto review/merge of PRs.

Has anyone tried this before? I suppose I could try to combine PRs for the past week and see if it works or if there are conflicts.

**Comment by naveensrinivasan** [comment]
> > we're on it!
> 
> Thanks Jory!!
> 
> I'll drop a few thoughts on things I'd like to see/have seen...
> 
> ### Multiple roles for a bot
> We have multiple bots in Kubernetes orgs. Most/all of them are user accounts; most/all run do their tasks with as few privileges on the org as possible:
> 
> ```yaml
>   bots:
>     description: Bot Service Accounts in the Kubernetes org
>     maintainers:
>     - k8s-ci-robot
>     - k8s-github-robot
>     - thelinuxfoundation
>     members:
>     - k8s-publishing-bot
>     - k8s-release-robot
>     privacy: closed
> ```
> 
> ref: https://github.com/kubernetes/org/blob/7b31d6ee782c82f5b44ec27600898271c292dbb4/config/kubernetes/org.yaml#L1425-L1434
> 
> In the case of `k8s-release-robot`, the account has multiple PATs; some have write access to repos, but most do not. Setting the correctly-scoped tokens as organization secrets (`SCORECARD_BOT_READ|WRITE_TOKEN`), so we can leverage on any scorecard repo should solve the problem.
> 
> ### Batching PRs / dependabot
> On the roadmap, but no timeline as of yet.
> 
> ref: https://twitter.com/stephenaugustus/status/1504611650443390983, [github/roadmap#148](https://github.com/github/roadmap/issues/148), [dependabot/dependabot-core#1190](https://github.com/dependabot/dependabot-core/issues/1190)
> 
> Batching is ideal, but approval is paramount. Any PR that merges should be subject to same review/checks, both from maintainers and presubmits.
> 
> An additional gate that we have in Kubernetes is requiring an `ok-to-test` label on PRs from "non-trusted" (non-org members) before we run presubmits.
> 
> Because dependabot is a GitHub App (and not an org member), we get around this limitation in SIG Release repos by assuming dependabot is trusted, and having it automatically apply the `ok-to-test` label for PRs: [kubernetes-sigs/release-utils#38](https://github.com/kubernetes-sigs/release-utils/pull/38)
> 
> Going a layer deeper on trust, certain repos/workflows can apply a `skip-review` label: https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.md#labels-that-apply-to-kubernetestest-infra-only-for-prs
> 
> This allows us to do things like automatically updating Prow images: [kubernetes/test-infra#25684](https://github.com/kubernetes/test-infra/pull/25684) Motivations and some detail on workflow for that here: https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/2539-continuously-deploy-k8s-prow
> 
> Back to dependabot...
> 
> In instances that our tests match our expectations, then a test failure says, "I, as a maintainer, do not trust this code to be accepted". This is currently true for any PR submitted.
> 
> For dependency updates, the human element of clicking `Approve` is mechanically and arguably, could be automated away.
> 
> Example workflow:
> 
> * @ossf-scorecard-bot is onboarded as a @ossf/scorecard-maintainers
> * bot scans for open PRs from dependabot that are not approved: https://github.com/ossf/scorecard/pulls?q=is%3Aopen+is%3Apr+author%3Aapp%2Fdependabot+review%3Arequired
> * bot approves the PR
> * --> auto-merge
> 
> Back to batching...
> 
> One of the things I crave in non-Kubernetes orgs is `tide` (the merge pool): https://github.com/kubernetes/test-infra/tree/master/prow/tide
> 
> We have a few expectations that are relevant here and enforced by branch protection:
> 
> * PR branches are up-to-date
> * no merge commits / require linear history
> 
> We violate merge commits on pull requests by merging `main` in to the PR branch, but that violation is nullified by the fact that we prefer squash and merge as a merge strategy.
> 
> A little out-of-scope here, but I have [Opinions ™️](https://twitter.com/stephenaugustus/status/1488662159869501442) and did a Twitter poll to check-in with other maintainers on the internet if you want to read more: https://twitter.com/stephenaugustus/status/1488624720400502790
> 
> tl;dr snippet:
> 
> > My answer is "it depends".
> > kubernetes/kubernetes does merge commits.
> > I've found with small projects/fewer maintainers, squashing works really nicely, especially if it's solving/circumventing teaching people to rebase.
> > As a more experienced maintainer, most of my commits are concise, focused on a specific area, and may include descriptions in the commit descriptions.
> > For a larger PR, you're going to lose the value of being able to interrogate those structured commits.
> > For a newer contributor, you may have a few scenarios:
> > 
> > * poorly-structured commit history
> > * commits which change multiple areas
> > * someone who does not yet know how to manage that proactively
> > 
> > Education is important here if you want to keep them contributing.
> 
> So in repos where multiple merge strategies are enabled, my preference is usually to do the following:
> 
> * `create a merge commit` if the submitter tells a good story with the commits and there's a higher value in having the individual commits stand on their own
> * `sqaush and merge` when the commit history is messy
> 
> I mention merge strategy because every dependabot PR leads to churn in multiple human approvals and in instances where the go.sum has been changed in a way that triggers another dependabot to start rebasing itself.
> 
> This goes away with a merge queue.
> 
> I haven't played around with it yet, but I would suggest a tool like Mergify: https://mergify.com/ (which is free for open source projects)
> 
> I've enabled it on a few of my orgs just now, but would be good to get some opinions from maintainers here.

Lots of information to read go down the rabbit hole 👍

**Comment by laurentsimon** [comment]
@[rarkins](https://github.com/rarkins) any plans to support batched pull requests in renovatebot? I found https://github.com/renovatebot/renovate/issues/4404#issuecomment-858554871 IIDC, this would allow us to batch PRs on a schedule?

**Comment by naveensrinivasan** [comment]
> > @strawberry-baked-alaska @brianbehlendorf -- Can you look into [1password for open source teams](https://github.com/1Password/1password-teams-open-source) for @ossf/scorecard-admins?
> 
> we're on it!

@jorydotcom Any updates on this?

**Comment by laurentsimon** [comment]
I think an answered question is whether we want to allow dependabot updates to be automatically merge-in or not.
For example it's sometimes possible to bypass reviews bu injecting code into dependabot PRs:
1. Dependabot publishes a change.
2. Maintainer edits that change to do whatever they want

Auto-merge happens.

I suggested we look into combining PRs in https://github.com/ossf/scorecard/issues/1729#issuecomment-1072945935, but have not spent time on it.

**Comment by naveensrinivasan** [comment]
> > > @strawberry-baked-alaska @brianbehlendorf -- Can you look into [1password for open source teams](https://github.com/1Password/1password-teams-open-source) for @ossf/scorecard-admins?
> > 
> > 
> > we're on it!
> 
> @jorydotcom Any updates on this?

@david-a-wheeler  @SecurityCRob Here is the follow-up on today's Scorecard meeting. 

We would like to start using the bot account.

**Comment by strawberry-baked-alaska** [comment]
Adding @theheels for assistance here!

**Comment by naveensrinivasan** [comment]
> Adding @theheels for assistance here!

Friendly ping.

**Comment by strawberry-baked-alaska** [comment]
@theheels Can you close this out or re-assign?
